### DISCUSSION TITLE
124p 3.7.3 3단계 부분!

### DISCUSSION THREAD

**Discussion by Youngju-Jang** [Study-2-Modern-Java-In-Action/Study-2-Modern-Java-in-Action#40] [CHAPTER 3 람다 표현식 - Q&A] [answered]
![IMG_6829](https://github.com/Study-2-Modern-Java-In-Action/Study-2-Modern-Java-in-Action/assets/77484084/fd67dfd7-c49e-4bd6-a77d-94679accb73b)

1.  아래 사진 설명글 부분 : 이해가 잘 안됨..
   ![image](https://github.com/Study-2-Modern-Java-In-Action/Study-2-Modern-Java-in-Action/assets/77484084/e957880e-14fd-49c1-a002-2b89d0fcfda0)

2.  기존에 sort안에서 구현시킨 Comparator의 메소드(`(a1, a2)-> a1.getWeight().compareTo(a2.getWeight()`)랑 
위에서 말하는 정적메소드 `comparing`은 다른메소드 인가요? 
    - 인수가 두개였다가 하나로바꼈는데 어떻게 되는건지?
    - 정적메소드란 무엇인가?

**Comment by Youngju-Jang** [comment]
<img width="737" alt="image" src="https://github.com/Study-2-Modern-Java-In-Action/Study-2-Modern-Java-in-Action/assets/77484084/3144b853-1558-41aa-aa6c-3023f2e4f416">

- 위에서 말하는 `정적메소드 comparing`이 이부분인것 같은데,  여기서 `(c1, c2)` 부분은 어디서 나온 값인가요?

**Reply by BvrPark** [reply]
저 부분은

Function 인터페이스
```java
@FunctionalInterface
public interface Function<T, R> {
    R apply(T t);
}
```

Comparator인터페이스의
```java
int compare(T o1, T o2);
```

메서드에서 나온 것으로 알고 있습니다!

**Comment by BvrPark** [comment, marked as answer]
### 질문 1.
우선 2가지 말씀드리자면,
1. Comparator 인터페이스에는 compareTo라는 메서드는 존재하지 않고 compare과 comparing과 같은 비교 메서드가 존재하는데 그 중, comparing이 정적 메서드 입니다.
2. compareTo는 Comparable이라는 인터페이스 안에 있는 메서드 입니다.

그래서
`Comparator는 Comparable 키를 추출해서 Comparator 객체로 만드는 Function 함수를 인수로 받는 정적 메서드 comparing을 포함한다.`라는 말은

```java
//Comparator 인터페이스 중 일부
//comparing 메서드
public static <T, U extends Comparable<? super U>> Comparator<T> comparing(
        Function<? super T, ? extends U> keyExtractor)
{
    Objects.requireNonNull(keyExtractor);
    return (Comparator<T> & Serializable)
        (c1, c2) -> keyExtractor.apply(c1).compareTo(keyExtractor.apply(c2));
}
```

```java
public static <T, U extends Comparable<? super U>> Comparator<T> comparing(
        Function<? super T, ? extends U> keyExtractor)
```
여기서 keyExtractor은 T타입의 객체를 입력 받아 U타입의 키를 반환하는 Function 함수인데 이것이 Comparable의 키를 추출하는 역할을 하는 것 같습니다.

</br>

```java
(Comparator<T> & Serializable) (c1, c2) -> keyExtractor.apply(c1).compareTo(keyExtractor.apply(c2))
```

* 여기는 Comparator 객체 구현을 반환합니다.
* keyExtractor.apply(c1), keyExtractor.apply(c2)를 Comparable 인터페이스의 compareTo를 사용해서 비교를 하는 것 같습니다.

</br>

그래서 제 생각은 comparing메서드를 구현해서 Function 함수로부터 키를 추출하고 그 추출한 키를 사용하여 객체를 비교하는 Comparator 객체를 생성한다는 말이 질문을 하셨던 말과 같은 것 같습니다.


### 질문 2
질문 2의 메인 질문은 질문 1에서 어느정도 답변이 된 것 같습니다.

* 인수가 두개였다가 하나로바꼈는데 어떻게 되는건지?
	* compareTo는 객체 2개를 인수로 받아서 비교하는 메서드이지만,
comparing은 객체 1개를 인수로 받아서 비교하는 정적 메서드이며 기본적으로 오름차순으로 정렬하는 Comparator을 생성합니다.
그래서
```java
//기존 코드
Comparator.comparing((Apple a -> a.getWeight())

// 메서드 참조 사용
Comparator.comparing(Apple::getWeight)
```
이런 식으로도 코드를 더 수정할 수도 있습니다.

* 정적메소드란 무엇인가?
정적 메서드는 객체 인스턴스에 관계없이 클래스 수준에서 호출되는 메서드를 말합니다.
```java
public class MethodTest {

    //정적 메서드
    public static int multi(int a, int b){
        return a * b;
    }

    //일반 메서드
    public int add(int a, int b){
        return a + b;
    }

    public static void main(String[] args) {
        MethodTest test = new MethodTest();
        test.add(1,2);
        multi(1,2); //이런 식으로 객체 인스턴스에 관계없이 바로 사용가능

    }
}
```

**Comment by JoisFe** [comment]
``` java
  public static <T, U extends Comparable<? super U>> Comparator<T> comparing(
            Function<? super T, ? extends U> keyExtractor)
    {
        Objects.requireNonNull(keyExtractor);
        return (Comparator<T> & Serializable)
            (c1, c2) -> keyExtractor.apply(c1).compareTo(keyExtractor.apply(c2));
    }
```
    
    질문의 의도가 c1, c2 2개가 어디서 나타났는지 의도인가요?
